[SPARK-14371] [MLLIB] OnlineLDAOptimizer should not collect stats for each doc in mini-batch to driver

akopich opened the pull request:
Hi, 

# What changes were proposed in this pull request?

as it was proposed by @jkbradley , ```gammat``` are not collected to the driver anymore. 

# How was this patch tested?
existing test suite.

akopich committed:
[SPARK-14371] [MLLIB] OnlineLDAOptimizer should not collect stats for each doc in mini-batch to driver

gammat are not collected to a local matrix but rather represented as RDD[BDV[Double]] and are aggregated in a distributed manner

SparkQA commented:
**[Test build #80524 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/80524/testReport)** for PR 18924 at commit [`f81f1cd`](https://github.com/apache/spark/commit/f81f1cdcf6de1dafdc79c1801cc2e2f1f803f4cc).

SparkQA commented:
**[Test build #80524 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/80524/testReport)** for PR 18924 at commit [`f81f1cd`](https://github.com/apache/spark/commit/f81f1cdcf6de1dafdc79c1801cc2e2f1f803f4cc).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/80524/
Test PASSed.

akopich commented:
@feynmanliang , @hhbyyh, @WeichenXu123, could you please review the PR?

WeichenXu123 commented:
Thanks! I will take a look later.

WeichenXu123 reviewed (commented):
@akopich I read your PR. It should be right. But I think the code here can be optimized better. We can merge computing `statsSum` and `logphat` into one pass, and eliminate persisting `stats` RDD. Like following code:

```
    val (statsSum: BDM[Double], logphat: BDV[Double]) = batch.mapPartitions { docs =>
      val nonEmptyDocs = docs.filter(_._2.numNonzeros > 0)

      val stat = BDM.zeros[Double](k, vocabSize)
      var logphatPart = BDV.zeros[Double](k)
      nonEmptyDocs.foreach { case (_, termCounts: Vector) =>
        val (gammad, sstats, ids) = OnlineLDAOptimizer.variationalTopicInference(
          termCounts, expElogbetaBc.value, alpha, gammaShape, k)
        stat(::, ids) := stat(::, ids).toDenseMatrix + sstats
        logphatPart += LDAUtils.dirichletExpectation(gammad)
      }
      Iterator((stat, logphatPart))
    }.treeAggregate((BDM.zeros[Double](k, vocabSize), BDV.zeros[Double](k)))(
      (u, v) => {u._1 += v._1; u._2 += v._2; u},
      (u, v) => {u._1 += v._1; u._2 += v._2; u}
    )
    logphat /= batchSize
```

Also cc @jkbradley

WeichenXu123 commented:
ping @akopich This is an very useful improvement. Can you update the code while you're at it ?

akopich commented:
Yes, sure. Thank you for the valuable comment. Hopefully, I'll update the code this week.

akopich committed:
[SPARK-14371] [MLLIB] OnlineLDAOptimizer should not collect stats for each doc in mini-batch to driver

In this commit I follow WeichenXu123's suggestion and merge computation of statsSum and logphat into a single pass.

Moreover, now logphat is computed only if optimizeDocConcentration is set to true, since it is not used otherwise.

akopich commented:
Ping @jkbradley .

Thank you @WeichenXu123 one again for the comment! Please, have a look.

SparkQA commented:
**[Test build #81885 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81885/testReport)** for PR 18924 at commit [`9ce9655`](https://github.com/apache/spark/commit/9ce9655d2275454e016dd3f3b640e578f4b6e0e4).

WeichenXu123 commented on mllib/src/main/scala/org/apache/spark/mllib/clustering/LDAOptimizer.scala:
```diff
@@ -462,31 +462,44 @@ final class OnlineLDAOptimizer extends LDAOptimizer {
+    logphatOption.foreach(updateAlpha(_, batchSize))
+
+    expElogbetaBc.destroy(false)
+    stats.unpersist()
```
This line is useless.

akopich replied:
Do you mean `stats.unpersist()`? Sure, I got rid of it.

SparkQA commented:
**[Test build #81885 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81885/testReport)** for PR 18924 at commit [`9ce9655`](https://github.com/apache/spark/commit/9ce9655d2275454e016dd3f3b640e578f4b6e0e4).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/81885/
Test FAILed.

WeichenXu123 commented on mllib/src/main/scala/org/apache/spark/mllib/clustering/LDAOptimizer.scala:
```diff
@@ -462,31 +462,44 @@ final class OnlineLDAOptimizer extends LDAOptimizer {
 
       val stat = BDM.zeros[Double](k, vocabSize)
-      var gammaPart = List[BDV[Double]]()
+      val logphatPartOption = logphatPartOptionBase
```
We can not reference outer variable and modify it in `map` function, it will generate undefined result.
You need create the `logphatPartOption` object in the `map` function, like:
```
val localOptimizeDocConcentration = optimizeDocConcentration
batch.mapPartitions { docs =>
  ...
  val logphatPartOption = if (localOptimizeDocConcentration) Some(BDV.zeros[Double](k)) else None
  ...
```
And note that avoid directly use `optimizeDocConcentration` in `rdd.map` function because the var  is class member and will cause the whole class object to serialize.

akopich replied:
Great point. Thank you.

WeichenXu123 reviewed (commented):
Thanks! There are still some issues.

akopich committed:
[SPARK-14371] [MLLIB] OnlineLDAOptimizer should not collect stats for each doc in mini-batch to driver

A bug caused by a reference to a local variable from map function is fixed.

akopich committed:
[SPARK-14371] [MLLIB] OnlineLDAOptimizer should not collect stats for each doc in mini-batch to driver

Useless unpersist removed.

SparkQA commented:
**[Test build #81893 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81893/testReport)** for PR 18924 at commit [`8e9bc73`](https://github.com/apache/spark/commit/8e9bc7308c0101131b6b1c6f7968c48514f18cce).

akopich commented:
@WeichenXu123, thank you for your prompt reply!

SparkQA commented:
**[Test build #81893 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81893/testReport)** for PR 18924 at commit [`8e9bc73`](https://github.com/apache/spark/commit/8e9bc7308c0101131b6b1c6f7968c48514f18cce).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/81893/
Test PASSed.

WeichenXu123 reviewed (commented):
Looks good, ping @jkbradley Can you take a look ?

jkbradley commented:
Taking a look

jkbradley commented on mllib/src/main/scala/org/apache/spark/mllib/clustering/LDAOptimizer.scala:
```diff
@@ -462,31 +462,46 @@ final class OnlineLDAOptimizer extends LDAOptimizer {
     val alpha = this.alpha.asBreeze
     val gammaShape = this.gammaShape
+    val optimizeDocConcentration = this.optimizeDocConcentration
+    val logphatPartOptionBase = () => if (optimizeDocConcentration) Some(BDV.zeros[Double](k))
```
Add an inline doc note here:
"We calculate logphat in the same pass as other statistics, but we only need it if we are optimizing docConcentration."

akopich replied:
OK.

jkbradley reviewed (commented):
Thanks a lot for this patch!  The math looks correct to me, though it's hard to verify via a test.

Have you tried checking to see if this change affects the results of the algorithm?  It'd be nice to do that with a quick manual test on a small dataset.

akopich commented on mllib/src/main/scala/org/apache/spark/mllib/clustering/LDAOptimizer.scala:
```diff
@@ -503,17 +518,15 @@ final class OnlineLDAOptimizer extends LDAOptimizer {
   /**
-   * Update alpha based on `gammat`, the inferred topic distributions for documents in the
-   * current mini-batch. Uses Newton-Rhapson method.
+   * Update alpha based on `logphat`.
```
I also rename `N` to `batchSize` which it is.

WeichenXu123 replied:
There're a small difference between old `N` and `batchSize`. `N` in old version code do not count non-empty docs.

akopich replied:
@WeichenXu123, you are right. So should we add `stats.count()` or should we rather embed the counting in the aggregation phase so that we avoid the second pass?

akopich replied:
Or another suggestion. Lets, have smth like 
`val stats: RDD[(BDM[Double], Option[BDV[Double]], Int)] = batch.mapPartitions {...}`  where the `Int` stands for the number of non-empty elements in a partition.  

What do you think?

WeichenXu123 replied:
I think it's OK. Let's keep the logic exactly the same with old version.

akopich replied:
But should we have `val stats: RDD[(BDM[Double], Option[BDV[Double]], Int)] = batch.mapPartitions {...}` stuff?

akopich replied:
Please, check out the updated PR. 

I have added `val stats: RDD[(BDM[Double], Option[BDV[Double]], Int)] = batch.mapPartitions {...}`.
Unfortunately, we cannot have the aggregation operation in a purely in-place manner now since `Int` is immutable. Shouldn't be a big deal since matrices and vectors are still updated in place.

WeichenXu123 replied:
I think it doesn't matter if `seq` do not return left param reference but return a new object.

akopich committed:
[SPARK-14371] [MLLIB] OnlineLDAOptimizer should not collect stats for each doc in mini-batch to driver

Comments added

SparkQA commented:
**[Test build #82007 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82007/testReport)** for PR 18924 at commit [`b3875e8`](https://github.com/apache/spark/commit/b3875e849a0659e9c2e5db16b97f909a0451ecd6).

akopich commented:
@jkbradley, thank you for your comments! Please, check out the commit adding the necessary docs.

Regarding tests: I believe, `OnlineLDAOptimizer alpha hyperparameter optimization` from `mllib/clustering/LDASuite.scala` covers the piece of code being rewritten. Or should there be tests of a different kind?

SparkQA commented:
**[Test build #82007 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82007/testReport)** for PR 18924 at commit [`b3875e8`](https://github.com/apache/spark/commit/b3875e849a0659e9c2e5db16b97f909a0451ecd6).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82007/
Test PASSed.

akopich committed:
[SPARK-14371] [MLLIB] OnlineLDAOptimizer should not collect stats for each doc in mini-batch to driver

Non-empty docs are properly counted now.

SparkQA commented:
**[Test build #82027 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82027/testReport)** for PR 18924 at commit [`8227b1a`](https://github.com/apache/spark/commit/8227b1af574e70aa3cc42f9b5146d5757e4cfb37).

akopich committed:
[SPARK-14371] [MLLIB] OnlineLDAOptimizer should not collect stats for each doc in mini-batch to driver

javadoc fix

SparkQA commented:
**[Test build #82030 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82030/testReport)** for PR 18924 at commit [`d7004a1`](https://github.com/apache/spark/commit/d7004a126a363e5bbfce5b335098f938f593517d).

SparkQA commented:
**[Test build #82027 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82027/testReport)** for PR 18924 at commit [`8227b1a`](https://github.com/apache/spark/commit/8227b1af574e70aa3cc42f9b5146d5757e4cfb37).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82027/
Test PASSed.

SparkQA commented:
**[Test build #82030 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82030/testReport)** for PR 18924 at commit [`d7004a1`](https://github.com/apache/spark/commit/d7004a126a363e5bbfce5b335098f938f593517d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82030/
Test PASSed.

jkbradley commented on mllib/src/main/scala/org/apache/spark/mllib/clustering/LDAOptimizer.scala:
```diff
@@ -462,36 +462,55 @@ final class OnlineLDAOptimizer extends LDAOptimizer {
+                                      else None
 
-    val stats: RDD[(BDM[Double], List[BDV[Double]])] = batch.mapPartitions { docs =>
+    val stats: RDD[(BDM[Double], Option[BDV[Double]], Int)] = batch.mapPartitions { docs =>
```
Let's use Long for the doc count since it could overflow for large datasets and miniBatchFraction

akopich replied:
Done.

jkbradley reviewed (commented):
Thanks for the updates!  I like the aggregation of the minibatch size within the same pass as the other stats.

I do wonder if we should use nonEmptyDocsN instead of batchSize within submitMiniBatch().  However, if we really want to handle things consistently, we should probably just filter out empty documents beforehand in order to make corpusSize, batchSize, and nonEmptyDocsN all refer to the same filtered corpus with all non-empty docs.  I'd suggest we do that in a follow-up JIRA & PR.

For the tests, I think it's fine for now to rely on the hyperparameter test comparing with gensim.

akopich committed:
[SPARK-14371] [MLLIB] OnlineLDAOptimizer should not collect stats for each doc in mini-batch to driver

Minibatch size is now stored in Long.

SparkQA commented:
**[Test build #82111 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82111/testReport)** for PR 18924 at commit [`1f9e650`](https://github.com/apache/spark/commit/1f9e650baab01c3b248d5b0225dbec1cde0c737d).

akopich commented:
@jkbradley, thanks for the comments. Who is supposed to create the followup jira?

SparkQA commented:
**[Test build #82111 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82111/testReport)** for PR 18924 at commit [`1f9e650`](https://github.com/apache/spark/commit/1f9e650baab01c3b248d5b0225dbec1cde0c737d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82111/
Test PASSed.

WeichenXu123 commented:
@akopich follow-up JIRA created here
https://issues.apache.org/jira/browse/SPARK-22111
Can you create follow up PR after this PR being merged ?

akopich commented:
@WeichenXu123, thanks for creating Jira. Yes, sure I will work on it.

akopich commented:
@WeichenXu123, @jkbradley, talking of merging. Is there anything else I should improve in this PR in order for it to be mergeable?

WeichenXu123 commented:
LGTM. Thanks! ping @jkbradley

akopich commented:
@WeichenXu123, the PR seems to receive no attention for 10 days now... What should I do?

WeichenXu123 commented:
Oh, sorry for that, it should waiting @jkbradley to merge it. Don't worry, I will contact him!

hhbyyh commented on mllib/src/main/scala/org/apache/spark/mllib/clustering/LDAOptimizer.scala:
```diff
@@ -462,36 +462,55 @@ final class OnlineLDAOptimizer extends LDAOptimizer {
 
   /**
-   * Update lambda based on the batch submitted. batchSize can be different for each iteration.
+   * Update lambda based on the batch submitted. nonEmptyDocsN can be different for each iteration.
```
comments should be consistent with code. Update code or revert comment change.

akopich replied:
Thanks. Comment reverted.

hhbyyh commented on mllib/src/main/scala/org/apache/spark/mllib/clustering/LDAOptimizer.scala:
```diff
@@ -462,36 +462,55 @@ final class OnlineLDAOptimizer extends LDAOptimizer {
+    val batchSize = (miniBatchFraction * corpusSize).ceil.toInt
+    updateLambda(batchResult, batchSize)
+
+    logphatOption.foreach(_ /= batchSize.toDouble)
```
Should use nonEmptyDocsN to be consistent with original implementation, also avoid divide by 0.

akopich replied:
Thanks for the good point. Do I understand correctly that if a batch without any non-empty docs is submitted, the `submitMiniBatch` method shouldn't change the state of `LDAOptimizer`? 
cc @WeichenXu123

jkbradley replied:
That sounds right to me.

hhbyyh replied:
agree.

hhbyyh reviewed (commented):
Sorry for the delayed response. Thanks for the great PR. 

IMO it's necessary to compare with the original implementation, especially when there're empty docs. The existing unit test may not be sufficient.

Also since the change targets performance improvement, a performance comparison is helpful. This can be done with some randomly generated data and together with the correctness verification.

akopich commented on mllib/src/main/scala/org/apache/spark/mllib/clustering/LDAOptimizer.scala:
```diff
@@ -462,36 +462,55 @@ final class OnlineLDAOptimizer extends LDAOptimizer {
     val expElogbetaBc = batch.sparkContext.broadcast(expElogbeta)
     val alpha = this.alpha.asBreeze
     val gammaShape = this.gammaShape
+    val optimizeDocConcentration = this.optimizeDocConcentration
```
This line is necessary in order to avoid serialization of `LDASuite` which is not serializable.

akopich commented on mllib/src/main/scala/org/apache/spark/mllib/clustering/LDAOptimizer.scala:
```diff
@@ -462,36 +462,55 @@ final class OnlineLDAOptimizer extends LDAOptimizer {
     val gammaShape = this.gammaShape
+    val optimizeDocConcentration = this.optimizeDocConcentration
+    // We calculate logphat in the same pass as other statistics, but we only need
+    // it if we are optimizing docConcentration
```
Thanks. Fixed.

akopich commented on mllib/src/main/scala/org/apache/spark/mllib/clustering/LDAOptimizer.scala:
```diff
@@ -462,36 +462,55 @@ final class OnlineLDAOptimizer extends LDAOptimizer {
+    // We calculate logphat in the same pass as other statistics, but we only need
+    // it if we are optimizing docConcentration
+    val logphatPartOptionBase = () => if (optimizeDocConcentration) Some(BDV.zeros[Double](k))
+                                      else None
```
Thanks. Fixed.

akopich commented on mllib/src/main/scala/org/apache/spark/mllib/clustering/LDAOptimizer.scala:
```diff
@@ -462,36 +462,55 @@ final class OnlineLDAOptimizer extends LDAOptimizer {
+    }
+
+    val elementWiseSum = (u : (BDM[Double], Option[BDV[Double]], Long),
+                                 v : (BDM[Double], Option[BDV[Double]], Long)) => {
```
Thanks. Fixed.

akopich commented on mllib/src/main/scala/org/apache/spark/mllib/clustering/LDAOptimizer.scala:
```diff
@@ -462,36 +462,55 @@ final class OnlineLDAOptimizer extends LDAOptimizer {
+      (u._1, u._2, u._3 + v._3)
+    }
+
+    val (statsSum: BDM[Double], logphatOption: Option[BDV[Double]], nonEmptyDocsN : Long) = stats
```
Thanks. Fixed.

akopich commented on mllib/src/main/scala/org/apache/spark/mllib/clustering/LDAOptimizer.scala:
```diff
@@ -462,36 +462,55 @@ final class OnlineLDAOptimizer extends LDAOptimizer {
     // Note that this is an optimization to avoid batch.count
-    updateLambda(batchResult, (miniBatchFraction * corpusSize).ceil.toInt)
-    if (optimizeDocConcentration) updateAlpha(gammat)
+    val batchSize = (miniBatchFraction * corpusSize).ceil.toInt
```
I believe, this will be settled down SPARK-22111.

hhbyyh replied:
this may wait.

hhbyyh replied:
Sure, since we're talking about consistency with old LDA. It's fine to keep using batchSize here.

akopich commented:
@WeichenXu123. thank you

akopich committed:
[SPARK-14371] [MLLIB] OnlineLDAOptimizer should not collect stats for each doc in mini-batch to driver

Comment and style fixes

akopich commented:
@hhbyyh, this change does not target performance but scalability, and I am afraid, the change is beneficial only for huge datasets and the tests would require massive computational resources.

SparkQA commented:
**[Test build #82448 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82448/testReport)** for PR 18924 at commit [`5c9547f`](https://github.com/apache/spark/commit/5c9547f6bc85b68d50b3c58a4a0d17eb6c75dcaf).

akopich commented:
BTW. Seems like `updateLambda` method relies (in older version as well) on `batchSize` only because this is `an optimization to avoid batch.count`. Shouldn't we rather use `nonEmptyDocsN` instead since we compute it efficietly now? But that is going to change logic...

SparkQA commented:
**[Test build #82448 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82448/testReport)** for PR 18924 at commit [`5c9547f`](https://github.com/apache/spark/commit/5c9547f6bc85b68d50b3c58a4a0d17eb6c75dcaf).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82448/
Test PASSed.

jkbradley commented on mllib/src/main/scala/org/apache/spark/mllib/clustering/LDAOptimizer.scala:
```diff
@@ -462,31 +462,54 @@ final class OnlineLDAOptimizer extends LDAOptimizer {
+    val batchSize = (miniBatchFraction * corpusSize).ceil.toInt
+    updateLambda(batchResult, batchSize)
+
+    logphatOption.foreach(_ /= nonEmptyDocsN.toDouble)
```
Good point about dividing by 0, @hhbyyh .  We should probably just check nonEmptyDocsN to see if it's 0, and if it is, skip all of these updates.  That's related to but actually separate from the follow-up SPARK-22111.

akopich replied:
Thanks for the comments, @jkbradley  and @hhbyyh. The check is added. I have also added a generation of warning message in case of an "empty" batch. I believe, a user should know that a thing like that happened.

jkbradley reviewed (commented):
Just a couple of comments.

@hhbyyh about testing this:
* Correctness: Are you recommending we take the original submitMiniBatch implementation and compare it with this new one?  I'd be OK with that.
* Scaling tests: I'm pretty unconcerned about this causing a regression since it does the same computation as before but distributes it, rather than collecting the data + computation to the driver.  @akopich perhaps you could run some local tests on a laptop just to check that the computational cost has not increased significantly.  That won't check communication, but communication should see a strict decrease from this fix.

akopich commented:
@jkbradley, thank you!

- Correctness: in order to test the equivalence of two versions of `submitMiniBatch` I have to bring both of them into the scope... One solution would be to derive a class `OldOnlineLDAOptimizer` from `OnlineLDAOptimizer` and override `submitMiniBatch` but the class is final. What's the preferred approach?
- Sure. Sounds good. Should I add test-case reporting the CPU time or should I rather define an `App`? Should I add the code to the PR or just report the results here?

hhbyyh commented:
Yes, I think local test is enough for both correctness and performance. 

For consistency with old LDA, just some manual local test would be sufficient. You may well just use the LDA example and switch the Spark jar files between Spark 2.2 and your branch. And I think the case with empty document worth special attention.

The same for the performance test. You may just post the result after your local test. I'm OK as long as there's no noticeable regression.

akopich commented:
Thank you, @hhbyyh.

I have augmented the example a bit: explicitly set random seed a nd chosen online optimizer:

`val lda = new LDA().setK(10).setMaxIter(10).setOptimizer("online").setSeed(13)`

But for some reason if I run it twice, the results are not the same. Is that expected? branch-2.2 was used.

akopich committed:
[SPARK-14371] [MLLIB] OnlineLDAOptimizer should not collect stats for each doc in mini-batch to driver

A check for the absence of non-empty documents in the batch is added. A warning message is issued in such a case.

SparkQA commented:
**[Test build #82482 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82482/testReport)** for PR 18924 at commit [`68ca837`](https://github.com/apache/spark/commit/68ca837569f9a0ae5cac8ced7f35b13e2f941522).

SparkQA commented:
**[Test build #82482 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82482/testReport)** for PR 18924 at commit [`68ca837`](https://github.com/apache/spark/commit/68ca837569f9a0ae5cac8ced7f35b13e2f941522).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `final class OnlineLDAOptimizer extends LDAOptimizer with Logging `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82482/
Test PASSed.

akopich commented:
I have conducted some performance testing with random data. 

The new implementation turns out to be notably faster. 

```
OLD with hyper-parameter optimization  : 237 sec
OLD w/o  hyper-parameter optimization  : 226 sec
NEW with hyper-parameter optimization : 178 sec
NEW w/o  hyper-parameter optimization : 171 sec
```

hhbyyh commented on mllib/src/main/scala/org/apache/spark/mllib/clustering/LDAOptimizer.scala:
```diff
@@ -462,31 +463,60 @@ final class OnlineLDAOptimizer extends LDAOptimizer {
+    // we calculate logphat in the same pass as other statistics.
+    // No calculation of loghat happens otherwise.
+    val logphatPartOptionBase = () => if (optimizeDocConcentration) {
+                                          Some(BDV.zeros[Double](k))
```
indent

akopich replied:
Thx. Fxd.

hhbyyh commented on mllib/src/main/scala/org/apache/spark/mllib/clustering/LDAOptimizer.scala:
```diff
@@ -462,31 +463,60 @@ final class OnlineLDAOptimizer extends LDAOptimizer {
+        elementWiseSum, elementWiseSum
+      )
+
+    if (nonEmptyDocsN == 0) {
```
I would use 

if (nonEmptyDocsN > 0) {
  update ...
} else {
  logWarning...
}
this

Just to avoid multiple exits.

akopich replied:
But spark scala style guide says :

"... \<when\> return is preferred: Use `return` as a guard to simplify control flow without adding a level of indentation".

hhbyyh replied:
I don't think that's the case here. But as long as all the cleanup work is done, I would not mind it.

hhbyyh commented on mllib/src/main/scala/org/apache/spark/mllib/clustering/LDAOptimizer.scala:
```diff
@@ -462,31 +463,60 @@ final class OnlineLDAOptimizer extends LDAOptimizer {
+    logphatOption.foreach(_ /= nonEmptyDocsN.toDouble)
+    logphatOption.foreach(updateAlpha(_, nonEmptyDocsN))
+
+    expElogbetaBc.destroy(false)
```
expElogbetaBc should always be destroyed.

akopich replied:
Great point. Thank you. Moreover, it should be destroyed as soon as there is no need in it anymore.

hhbyyh reviewed (commented):
I did some investigation and found OnlineLDAOptimizer. variationalTopicInference did not use the random seed. 

I created an issue https://issues.apache.org/jira/browse/SPARK-22210 for it. Feel free to resolve it in this or other PR.

Given the inconsistent LDA output, I'm OK to bypass the consistency check with old implementation.

akopich commented on mllib/src/main/scala/org/apache/spark/mllib/clustering/LDAOptimizer.scala:
```diff
@@ -462,31 +463,60 @@ final class OnlineLDAOptimizer extends LDAOptimizer {
+    val optimizeDocConcentration = this.optimizeDocConcentration
+    // If and only if optimizeDocConcentration is set true,
+    // we calculate logphat in the same pass as other statistics.
+    // No calculation of loghat happens otherwise.
```
If I don't assign logphatPartOptionBase to a local variable, NonSerializableException is generated.

Regarding comments. Isn't it necessary to emphasise that the computation happens in the same pass?

hhbyyh replied:
Maybe define logphatPartOptionBase as Option but not function.
val logphatPartOptionBase = if (optimizeDocConcentration) {
                                         Some(BDV.zeros[Double](k))

hhbyyh replied:
About the comments. Keep it as you wish.

akopich replied:
About `logphatPartOptionBase`: tried that, initially and failed. This was discussed above with @WeichenXu123. The problem is caused by in-place modifications.

hhbyyh replied:
I see.

akopich committed:
[SPARK-14371] [MLLIB] OnlineLDAOptimizer should not collect stats for each doc in mini-batch to driver

`expElogbetaBc` is destroyed anyway now. Also indentation is fixed.

SparkQA commented:
**[Test build #82487 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82487/testReport)** for PR 18924 at commit [`2942082`](https://github.com/apache/spark/commit/294208217b0cbf0fe745b7a8c603ec2f5675f5dc).

SparkQA commented:
**[Test build #82487 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82487/testReport)** for PR 18924 at commit [`2942082`](https://github.com/apache/spark/commit/294208217b0cbf0fe745b7a8c603ec2f5675f5dc).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82487/
Test PASSed.

hhbyyh commented on mllib/src/main/scala/org/apache/spark/mllib/clustering/LDAOptimizer.scala:
```diff
@@ -462,31 +463,60 @@ final class OnlineLDAOptimizer extends LDAOptimizer {
+    // If and only if optimizeDocConcentration is set true,
+    // we calculate logphat in the same pass as other statistics.
+    // No calculation of loghat happens otherwise.
+    val logphatPartOptionBase = () => if (optimizeDocConcentration) {
```
Not an style expert myself. Just what I would use:
```
    val logphatPartOptionBase = () => {
      if (optimizeDocConcentration) Some(BDV.zeros[Double](k)) else None
    }
```

hhbyyh commented on mllib/src/main/scala/org/apache/spark/mllib/clustering/LDAOptimizer.scala:
```diff
@@ -462,31 +463,60 @@ final class OnlineLDAOptimizer extends LDAOptimizer {
+    }
+
+    val elementWiseSum = (u : (BDM[Double], Option[BDV[Double]], Long),
+                          v : (BDM[Double], Option[BDV[Double]], Long)) => {
```
Minor: indent.

akopich replied:
Do you mean the extra spaces after `u` and `v`?

hhbyyh replied:
```
val elementWiseSum = (u: (BDM[Double], Option[BDV[Double]], Long),
    v: (BDM[Double], Option[BDV[Double]], Long)) => {
```

akopich replied:
I see now. Thank you. But seems like the style guide suggests to move both of the parameters to the new line.

hhbyyh commented on mllib/src/main/scala/org/apache/spark/mllib/clustering/LDAOptimizer.scala:
```diff
@@ -503,21 +533,22 @@ final class OnlineLDAOptimizer extends LDAOptimizer {
+   * @param nonEmptyDocsN number of non-empty documents
    */
-  private def updateAlpha(gammat: BDM[Double]): Unit = {
+  private def updateAlpha(logphat: BDV[Double], nonEmptyDocsN : Double): Unit = {
```
nonEmptyDocsN: Int ?

akopich replied:
The methods will have to cast `nonEmptyDocsN: Int` to `Double`. This way we have the conversion implicitly, but the method is private so I don't think it's going to hurt.

hhbyyh reviewed (approved):
Only some debatable style issue. I'm done here. LGTM

akopich committed:
[SPARK-14371] [MLLIB] OnlineLDAOptimizer should not collect stats for each doc in mini-batch to driver

Indentation fix.

SparkQA commented:
**[Test build #82505 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82505/testReport)** for PR 18924 at commit [`f181496`](https://github.com/apache/spark/commit/f1814965885e0c82a71287f5e5912e11b126b8a4).

akopich committed:
[SPARK-14371] [MLLIB] OnlineLDAOptimizer should not collect stats for each doc in mini-batch to driver

Extra spaces before semicolons are removed (like `val a : Int = 1`).

SparkQA commented:
**[Test build #82506 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82506/testReport)** for PR 18924 at commit [`a81dae5`](https://github.com/apache/spark/commit/a81dae574f2085ec390effd1b9b1962970f00239).

akopich commented:
So shall we ping @jkbradley, shan't we?

SparkQA commented:
**[Test build #82505 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82505/testReport)** for PR 18924 at commit [`f181496`](https://github.com/apache/spark/commit/f1814965885e0c82a71287f5e5912e11b126b8a4).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82505/
Test PASSed.

SparkQA commented:
**[Test build #82506 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82506/testReport)** for PR 18924 at commit [`a81dae5`](https://github.com/apache/spark/commit/a81dae574f2085ec390effd1b9b1962970f00239).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82506/
Test PASSed.

akopich commented:
@WeichenXu123, could you please notify @jkbradley once again?

WeichenXu123 commented:
@akopich  LGTM. and do you have time to create a PR to resolve random seed not working issue mentioned by @hhbyyh ?  Thanks!

akopich commented:
@WeichenXu123, yes sure. But can this wait until this PR is merged?

WeichenXu123 commented:
@akopich Yes you can wait this to be merged first. I think @jkbradley will have time to check this next week. Don't worry!

akopich commented:
@WeichenXu123, no problem! Thank you.

jkbradley commented:
LGTM
Sorry for the delay!
I'll merge it after re-running tests

SparkQA commented:
**[Test build #3951 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3951/testReport)** for PR 18924 at commit [`a81dae5`](https://github.com/apache/spark/commit/a81dae574f2085ec390effd1b9b1962970f00239).

akopich commented:
@jkbradley, no problem. The test build seems to be aborted. What's wrong?

akopich commented:
retest this please

SparkQA commented:
**[Test build #82880 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82880/testReport)** for PR 18924 at commit [`a81dae5`](https://github.com/apache/spark/commit/a81dae574f2085ec390effd1b9b1962970f00239).

SparkQA commented:
**[Test build #82880 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82880/testReport)** for PR 18924 at commit [`a81dae5`](https://github.com/apache/spark/commit/a81dae574f2085ec390effd1b9b1962970f00239).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82880/
Test PASSed.

akopich commented:
ping @jkbradley. Anyway, tests are passed now.

jkbradley commented:
Merging with master

jkbradley commented:
@akopich I'm afraid pings on Git don't work for me; I just have too many to keep up with.  Again, sorry for the delays; I have very limited bandwidth nowadays.

jkbradley commented:
I'll update JIRA later; it seems like Apache JIRA is having problems right now.

akopich commented:
@jkbradley, no problem.
@jkbradley, @WeichenXu123, @hhbyyh, thank you all guys!
